KIPs/kip-113.md

```diff
@@ -0,0 +1,216 @@
+
+## Abstract
+
+This standard defines an interface for BLS12-381 registry, which enables nodes or smart contracts to access and view all the public keys of GCs. This is not to replace AddressBook, but rather to supplement it in order to facilitate the use of BLS in various components, including signature-based random (KIP-114).
+
+## Motivation
+
+The current design of AddressBook cannot incorporate BLS12-381 keys. The lack of BLS registry limits the use of BLS from the node. The objective is to design a simple registry that includes essential functions, and minimal access control, thereby reducing any potential security risks.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+Registry contract must implement the IKIP_113 interface:
+
+```solidity
+pragma solidity ^0.8.19;
+
+/// @title KIP-113 BLS public key registry
+/// @dev See https://github.com/klaytn/kips/issues/113
+interface IKIP_113 {
+    struct BlsPublicKeyInfo {
+        // compressed BLS12-381 public key (48 bytes)
+        bytes publicKey;
+
+        // proof-of-possession (96 bytes)
+        // must be a result of PopProve algorithm as per
+        // draft-irtf-cfrg-bls-signature-05 section 3.3.3.
+        // with ciphersuite "BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_"
+        bytes pop;
+    }
+
+    event PubkeyRegistered(address cnNodeId, bytes publicKey, bytes pop);
+    event PubkeyUnegistered(address cnNodeId, bytes publicKey, bytes pop);
+
+    function registerPublicKey(bytes memory publicKey, bytes memory pop) external;
+    function unregisterPublicKey(address cnNodeId) external;
+
+    function getInfo(address cnNodeId) external view returns (BlsPublicKeyInfo pubkey);
+    function getAllInfo() external view returns (address[] memory nodeIdList, BlsPublicKeyInfo[] memory pubkeyList);
+}
+```
+
+### Definitions
+
+- Public key: BLS12-381 key under [minimal-pubkey-size](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#section-2.1-2.2.2) configuration (48 bytes)
+- Proof-of-possession (pop): the signature of the hash of the public key (96 bytes)
+
+### Smart Contracts Overview
+
+The proposed smart contract will be implemented in Solidity and will be compatible with the Ethereum Virtual Machine (EVM).
+
+The smart contract will have the following features:
+
+- Register a public key and a proof-of-possession
+- Unregister a public key
+- View the registered public keys
+
+#### Methods
+
+##### registerPublicKey
+
+Registers the given public key associated with the `msg.sender` address.
+
+The function validates the following requirements:
+
+- The function MUST revert if `publicKey.length != 48` or `pop.length != 96`
+- The function MUST revert if `publicKey` or `pop` is equal to zero
+- The function SHOULD authenticate and authorize `msg.sender`
```

Comment:
Can you more elaborate what kinds of authentication and authorization will be done in this function?

Reply:
It's chain-specific so I wrote it in the below "cypress-specific implementation" section. It's not mandatory so service chain operators can decide to not perform authentication/authorization.

Reply:
Yeah, but it is not easy to be linked the things. What about adding some more descriptions like:

```suggestion
- The function SHOULD authenticate and authorize `msg.sender`. The implementation can be different from chain to chain. For Cypress (Klaytn's mainnet), please refer to the section `Cypress-specific implementation`.
```

Reply:
Added in dadd0d6